File: DSA/Pragya_Agarwal/Day13_Question1.py
def word_break(s,Dict):
    word_set= set(Dict)
    n = len(s)
    dp = [False] * (n + 1)
    dp[0] = True
    
    for i in range(1, n + 1):
        for j in range(i):
            if dp[j] and s[j:i] in word_set:
                dp[i] = True
                break
    return dp[n]
    
string = "applepie"
dictionary = ["apple","pie"]
if word_break(string, dictionary):
   print("Yes")
else:
   print("No")
  

File: DSA/Arshdeep-Kaur/Day13_Question1.cpp
#include <iostream>
#include <vector>
#include <string>
using namespace std;

bool isInDictionary(const string &word, const vector<string> &dictionary) {
    for (const string &dictWord : dictionary) {
        if (word == dictWord) {
            return true;
        }
    }
    return false;
}

bool canSegmentString(const string &s, const vector<string> &dictionary) {
    int n = s.length();
    vector<bool> dp(n + 1, false);
    dp[0] = true;

    for (int i = 1; i <= n; ++i) {
        for (int j = 0; j < i; ++j) {
            if (dp[j] && isInDictionary(s.substr(j, i - j), dictionary)) {
                dp[i] = true;
                break;
            }
        }
    }

    return dp[n];
}

int main() {
    string s = "applepie";
    vector<string> dictionary = {"apple", "pie"};

    if (canSegmentString(s, dictionary)) {
        cout << "YES" << endl;
    } else {
        cout << "NO" << endl;
    }
}
